### PR TITLE
[FEATURE] Added scraper to add all translations to the database.

### PR DESCRIPTION
**Pull request by borisvankatwijk**
This makes it possible for merchant/clients to translate any string in the adminpanel of Magento.

- Added Console command to scrape all translations
    --Option of including missing translation files if they are gathered (if file exists)
- Added crontab to insert new translations nightly at 3:13 AM (i.e. if new module is installed or translatable strings changed)
- Minor UX / syntax improvements in loose commits

### CODE REVIEW

**Comment by borisvankatwijk** [comment]
Missing features (to-do's next release)

Add flag to translation database; if user edit's a translation, user_defined is flagged as true.
Add --force update for every entry that is not flagged as user_defined; to update csv changes into database.

File: view/adminhtml/ui_component/experius_translation_index.xml [hunk 1]
```diff
@@ -58,7 +52,7 @@
 			<item name="config" xsi:type="array">
 				<item name="editorConfig" xsi:type="array">
 					<item name="selectProvider" xsi:type="string">experius_translation_index.experius_translation_index.experius_translation_columns.ids</item>
-					<item name="enabled" xsi:type="boolean">false</item>
+					<item name="enabled" xsi:type="boolean">true</item>
```
Comment (lewisvoncken) [inline view/adminhtml/ui_component/experius_translation_index.xml]: @borisvankatwijk with inline edit the value of Has a translation hasn't been updated, oh it also doesn't work in edit view.

File: Setup/UpgradeSchema.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    ) {
+        $setup->startSetup();
+        if (version_compare($context->getVersion(), "1.1.0", "<")) {
+            $setup->getConnection()->addColumn(
+                $setup->getTable('translation'),
+                'different',
+                [
+                    'type' => \Magento\Framework\DB\Ddl\Table::TYPE_SMALLINT,
+                    'nullable' => false,
+                    'default' => 0,
```
[the remark this answers is not in the document]
Comment (lewisvoncken) [reply Setup/UpgradeSchema.php]: @borisvankatwijk this field is only set with the import to database and not when the string is edited in the admin

File: view/adminhtml/ui_component/experius_translation_index.xml [hunk 3]
```diff
@@ -116,11 +99,41 @@
...
 					<item name="filter" xsi:type="string">text</item>
 					<item name="label" translate="true" xsi:type="string">Translated String</item>
 				</item>
 			</argument>
 		</column>
+		<column name="locale">
+			<argument name="data" xsi:type="array">
+				<item name="options" xsi:type="object">Experius\MissingTranslations\Model\Config\Source\Locale</item>
+				<item name="config" xsi:type="array">
+					<item name="filter" xsi:type="string">select</item>
```
Comment (lewisvoncken) [inline view/adminhtml/ui_component/experius_translation_index.xml]: @borisvankatwijk please edit pre filter to so when you go back to the grid overview it is filtered with the previously set filters.
for more information see the PageNotFound module https://github.com/experius/Magento-2-Module-PageNotFound/blob/master/view/adminhtml/ui_component/experius_page_not_found_index.xml#L45

maybe it should be bookmark and not bookmarks

File: etc/crontab.xml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0"?>
+<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Cron:etc/crontab.xsd">
+    <group id="default">
+        <job name="update_all_translations_to_database" instance="Experius\MissingTranslations\Cron\Collect" method="execute">
+            <schedule>13 3 * * *</schedule>
```
[the remark this answers is not in the document]
Comment (lewisvoncken) [reply etc/crontab.xml]: @borisvankatwijk I would prefer two crons so it is clear that one does the existing and the other the missing and by default these are disable through a configuration setting. I understand the need of the functionality but the module is for MissingTranslations so I think in default it only should run for missing translations to import and when you configure it should also run for existing ones.

[the remark this answers is not in the document]
Comment (borisvankatwijk) [reply etc/crontab.xml]: Hi @lewisvoncken ,

Would it be beneficial to split the modules? I thought about a rename to \Experius\Translation, which has missingtranslations as a part of the module. But this namespace was already claimed and seemed like redundant work.

Timeslot was chosen due to many heavy processes having the whole hour timeslots (0 3), and therefore this timeslot has less chance of causing issue's within any shop.

File: Console/Command/AddTranslationsToDatabaseCommand.php [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+            if ($input->getOption(self::INPUT_KEY_STORE)) {
+                throw new \InvalidArgumentException('Store is not needed when --global flag is set.');
+            }
+        } elseif (!$input->getOption(self::INPUT_KEY_STORE)) {
+            throw new \InvalidArgumentException('Store is needed when --global flag is not set.');
+        }
+
+        $storeId = $global ? '0' : $input->getOption(self::INPUT_KEY_STORE);
+
+        $includeMissing = false;
```
[the remark this answers is not in the document]
Comment (borisvankatwijk) [reply Console/Command/AddTranslationsToDatabaseCommand.php]: Hi @lewisvoncken ,

That would be the best solution, I agree.

Would it be smart for a next release to remove the "filebase" dump to csv, and Include these new commands?
experius_missingtranslations:process-existing
experius_missingtranslations:process-missing

which add them to the database only.

File: view/adminhtml/ui_component/experius_translation_index.xml (hunk 1 above)
Comment (borisvankatwijk) [reply view/adminhtml/ui_component/experius_translation_index.xml]: Note to myself; @TODO: Add event / plugin to update "has a translation" on translation save.

File: view/adminhtml/ui_component/experius_translation_index.xml (hunk 3 above)
Comment (borisvankatwijk) [reply view/adminhtml/ui_component/experius_translation_index.xml]: Note to myself; @TODO: Add bookmark function

File: Ui/Component/Listing/Column/TranslationActions.php [hunk 6]
```diff
@@ -57,19 +57,6 @@ public function prepareDataSource(array $dataSource)
                                 ]
                             ),
                             'label' => __('Edit')
-                        ],
```
[the remark this answers is not in the document]
Comment (borisvankatwijk) [reply Ui/Component/Listing/Column/TranslationActions.php]: Will be re-added

File: view/adminhtml/ui_component/experius_translation_index.xml (hunk 1 above)
Comment (borisvankatwijk) [reply view/adminhtml/ui_component/experius_translation_index.xml]: @lewisvoncken ,

Has been added

File: view/adminhtml/ui_component/experius_translation_index.xml (hunk 3 above)
Comment (borisvankatwijk) [reply view/adminhtml/ui_component/experius_translation_index.xml]: @lewisvoncken ,

"bookmarks" was correct, has been fixed

**Comment by borisvankatwijk** [comment]
FIXED ui component -> bookmarks
FIXED no inline delete but inline edit
REMOVED (redundant) shortkey for include is not working (but can be solved by splitting the command)
ADDED has a translation doesn't get updated after save
ADDED has a translation doesn't get set to the right value when updating module in existing envirnoment
ADDED split the cron to two crons for existing and missing translations
ADDED split the console command to two commands for existing and missing translations
TODO add configuration to by default disable the cron for existing ones because the main feature is missing translations

**Comment by borisvankatwijk** [comment]
Added latest bugfix, which prevents admin from changing locale on save.

**Comment by lewisvoncken** [comment]
@borisvankatwijk 
as an answer on your question:
How do you propose to disable the cron be default? Is this simply a parameter one can give with the crontab xml?

I recommend to just add an enabled check in the cron collect class

https://github.com/experius/Magento-2-Module-Experius-MissingTranslations/blob/feature-database/Cron/Collect.php

**Comment by borisvankatwijk** [comment]
Good point, will look into it. Thought maybe there was a xml setting instead of adding a system.xml config.
To be continue'd =D

**Pull request merged**